### ISSUE TITLE
ShallowCopy for byte::Bytes

### ISSUE THREAD

**Issue by PlaneB** [jonhoo/rust-evmap#16]
Seems like byte::Bytes can not be used as type, which is unfortunate, as it should be more or less among the most favored types as it is made for fast data transfer, continuous memory footprint and such.

copy in Bytes is shallow by design, so the following compiler error might be avoidable...

error[E0277]: the trait bound `bytes::Bytes: evmap::ShallowCopy` is not satisfied
   --> src/foo.rs:143:5
    |
143 |     w:WriteHandle<[u8; 32], Bytes, (), BuildNoHashHasher<u32>>,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `evmap::ShallowCopy` is not implemented for `bytes::Bytes`
    |
    = note: required by `evmap::WriteHandle`

error: aborting due to previous error

**Comment by jonhoo** [comment]
Hmm.. I _think_ it should be okay do have:
```rust
impl ShallowCopy for Bytes {
    unsafe fn shallow_copy(&mut self) -> Self {
        Bytes::from_static(self.as_ref())
    }
}
```
Along those lines, I _also_ think it would be good to add
```rust
impl<'a, T> ShallowCopy for &'a [T] {
    unsafe fn shallow_copy(&mut self) -> Self {
        std::slice::from_raw_parts(self.as_ptr(), self.len())
    }
}
```
I'd be happy to take a look at a PR that adds these :)

**Comment by PlaneB** [comment]
Sorry 'PR' ?

**Comment by PlaneB** [comment]
You want to look into a program that uses those trait implementations? I will try them eventually. What minimal code do you need and why you want to see it?

**Comment by jonhoo** [comment]
I'm not sure I understand what you mean? Those trait implementations would need to be added to `evmap` itself for you to be able to rely on them in an external crate.

**Comment by PlaneB** [comment]
A really? Ok I was under the impression, that the user software implement those traits for Bytes in their code.

**Comment by jonhoo** [comment]
Ah, no, we're limited here by Rust's [_orphan rule_](https://doc.rust-lang.org/book/ch10-02-traits.html#implementing-a-trait-on-a-type):

> One restriction to note with trait implementations is that we can implement a
> trait on a type only if either the trait or the type is local to our crate.
> For example, we can implement standard library traits like `Display` on a
> custom type like `Tweet` as part of our aggregator crate functionality,
> because the type `Tweet` is local to our aggregator crate. We can also
> implement `Summary` on `Vec<T>` in our aggregator crate, because the trait
> `Summary` is local to our aggregator crate.
> 
> **But we can’t implement external traits on external types.** For example, we
> can’t implement the `Display` trait on `Vec<T>` within our aggregator crate,
> because `Display` and `Vec<T>` are defined in the standard library and aren’t
> local to our aggregator crate. This restriction is part of a property of
> programs called _coherence_, and more specifically the _orphan rule_, so
> named because the parent type is not present. This rule ensures that other
> people’s code can’t break your code and vice versa. Without the rule, two
> crates could implement the same trait for the same type, and Rust wouldn’t
> know which implementation to use.

In other words, since `evmap` defines the `ShallowCopy` trait, _either_ the `byte` crate would have to implement `ShallowCopy` for `Bytes` (which seems unlikely) _or_ that implementation has to exist inside of `evmap`. Since `byte::Bytes` is so common, I think I'd be okay with including that impl directly here :)

**Comment by PlaneB** [comment]
I see.

**Comment by jonhoo** [comment]
Or you can [submit a pull request](https://help.github.com/en/articles/creating-a-pull-request) yourself that makes the necessary changes (adding a dependency on `byte`, the `impl`s above, and tests with the corresponding types) which I could then review and merge :)

**Comment by jonhoo** [comment]
Fixed in #30!
